src/circuit_builder.rs

```diff
@@ -0,0 +1,304 @@
+//! Tools & traits for PLONK circuits
+
+use crate::commitment_scheme::kzg10::PublicParameters;
+use crate::constraint_system::StandardComposer;
+use crate::proof_system::{Proof, ProverKey, VerifierKey};
+use anyhow::{Error, Result};
+use dusk_bls12_381::Scalar as BlsScalar;
+use dusk_jubjub::{AffinePoint as JubJubAffine, Scalar as JubJubScalar};
+
+/// Circuit inputs
+#[derive(Debug, Clone, Copy)]
+pub struct CircuitInputs<'a> {
+    bls_scalars: &'a [BlsScalar],
+    jubjub_scalars: &'a [JubJubScalar],
+    jubjub_affines: &'a [JubJubAffine],
+}
+
+/// Public Input
+#[derive(Debug, Copy, Clone)]
+pub enum PublicInput {
+    /// Scalar Input
+    BlsScalar(BlsScalar, usize),
+    /// Embedded Scalar Input
+    JubJubScalar(JubJubScalar, usize),
+    /// Point as Public Input
+    AffinePoint(JubJubAffine, usize, usize),
+}
+
+impl PublicInput {
+    #[allow(dead_code)]
+    fn value(&self) -> Vec<BlsScalar> {
+        match self {
+            PublicInput::BlsScalar(scalar, _) => vec![*scalar],
+            PublicInput::JubJubScalar(scalar, _) => vec![BlsScalar::from(*scalar)],
+            PublicInput::AffinePoint(point, _, _) => {
+                vec![point.get_x().into(), point.get_y().into()]
+            }
+        }
+    }
+}
+
+/// Circuit representation for a gadget with all of the tools that it
+/// should implement.
+pub trait Circuit<'a>
+where
+    Self: Sized + Default,
+{
+    /// Gadget implementation used to fill the composer.
+    fn gadget(
+        composer: &mut StandardComposer,
+        inputs: CircuitInputs,
+    ) -> Result<(Vec<PublicInput>, usize), Error>;
```

Comment:
I think we should return only the `[PublicInput]` since the circuit size can be trivially obtained by the consumer via the mutated composer.

This way we simplify the documentation and the signature

Reply:
The idea is that you, as a user never interact with the composer at all. So therefore we need to return this.